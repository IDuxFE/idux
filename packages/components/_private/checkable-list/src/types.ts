/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VirtualScrollToFn } from '@idux/cdk/scroll'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface CheckableListData {
  key?: VKey
  label?: string
  disabled?: boolean
  [key: string]: unknown
}

export interface CheckableListScroll {
  height?: string | number
  fullHeight?: boolean
}

export const checkableListProps = {
  dataSource: Array as PropType<CheckableListData[]>,
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  checked: Function as PropType<(item: CheckableListData) => boolean>,
  customAdditional: { type: Function as PropType<CheckableListCustomAdditional>, default: undefined },
  disabled: Function as PropType<(item: CheckableListData) => boolean>,
  getKey: Function as PropType<(item: CheckableListData) => VKey>,
  labelKey: String,
  virtual: {
    type: Boolean,
    default: false,
  },
  scroll: Object as PropType<CheckableListScroll>,
  onCheckChange: [Function, Array] as PropType<MaybeArray<(item: CheckableListData, checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<(item: CheckableListData) => void>>,
  onScroll: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onScrolledChange: [Function, Array] as PropType<
    MaybeArray<(startIndex: number, endIndex: number, visibleData: unknown[]) => void>
  >,
  onScrolledBottom: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export const checkableListItemProps = {
  checked: {
    type: Boolean,
    default: false,
  },
  checkable: {
    type: Boolean,
    default: true,
  },
  removable: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  label: String,
  value: {
    type: [String, Number, Symbol] as PropType<VKey>,
    required: true,
  },
  onCheckChange: [Function, Array] as PropType<MaybeArray<(checked: boolean) => void>>,
  onRemove: [Function, Array] as PropType<MaybeArray<() => void>>,
} as const

export interface CheckableListApi {
  scrollTo: VirtualScrollToFn
}

export type CheckableListProps = ExtractInnerPropTypes<typeof checkableListProps>
export type CheckableListItemProps = ExtractInnerPropTypes<typeof checkableListItemProps>
export type CheckableListPublicProps = ExtractPublicPropTypes<typeof checkableListProps>
export type CheckableListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CheckableListPublicProps> & CheckableListPublicProps,
  CheckableListApi
>
export type CheckableListInstance = InstanceType<DefineComponent<CheckableListProps, CheckableListApi>>

export type CheckableListCustomAdditional = (options: {
  data: CheckableListData
  index: number
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Record<string, any> | undefined
