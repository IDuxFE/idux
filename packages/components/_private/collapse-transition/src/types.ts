/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export type CollapseTransitionMode = 'height' | 'width'

export const collapseTransitionProps = {
  appear: IxPropTypes.bool.def(false),
  name: IxPropTypes.string,
  mode: IxPropTypes.oneOf<CollapseTransitionMode>(['height', 'width']).def('height'),
  onAfterEnter: IxPropTypes.emit(),
  onAfterLeave: IxPropTypes.emit(),
}

export type CollapseTransitionProps = ExtractInnerPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionPublicProps = ExtractPublicPropTypes<typeof collapseTransitionProps>
export type CollapseTransitionComponent = DefineComponent<
  Omit<HTMLAttributes, keyof CollapseTransitionPublicProps> & CollapseTransitionPublicProps
>
export type CollapseTransitionInstance = InstanceType<DefineComponent<CollapseTransitionProps>>
