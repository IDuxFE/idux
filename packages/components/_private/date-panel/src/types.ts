/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export const datePanelProps = {
  cellTooltip: Function as PropType<(cell: { value: Date; disabled: boolean }) => string | void>,
  disabledDate: Function as PropType<(date: Date) => boolean>,
  type: {
    type: String as PropType<DatePanelType>,
    default: 'date',
  },
  value: [Date, Array] as PropType<Date | (Date | undefined)[]>,
  activeDate: Date,
  visible: {
    type: Boolean,
    default: undefined,
  },
  isSelecting: {
    type: Boolean,
    default: undefined,
  },

  onCellClick: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  onCellMouseenter: [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
  'onUpdate:activeDate': [Function, Array] as PropType<MaybeArray<(date: Date) => void>>,
}

export type DatePanelProps = ExtractInnerPropTypes<typeof datePanelProps>
export type DatePanelPublicProps = ExtractPublicPropTypes<typeof datePanelProps>
export type DatePanelComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DatePanelPublicProps> & DatePanelPublicProps
>
export type DatePanelInstance = InstanceType<DefineComponent<DatePanelProps>>

export type DatePanelType = 'date' | 'week' | 'month' | 'quarter' | 'year'

// private
export const panelRowProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
}

export const panelCellProps = {
  rowIndex: {
    type: Number,
    required: true,
  },
  cellIndex: {
    type: Number,
    required: true,
  },
}
