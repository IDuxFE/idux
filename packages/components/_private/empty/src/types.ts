/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { EmptyProps as IxEmptyProps } from '@idux/components/empty'

export interface EmptyProps {
  empty?: string | IxEmptyProps
}
