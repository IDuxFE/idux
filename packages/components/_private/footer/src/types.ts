/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ButtonProps } from '@idux/components/button'
import type { DefineComponent, HTMLAttributes, PropType, VNode } from 'vue'

export interface FooterButtonProps extends ButtonProps {
  key?: VKey
  text?: string | VNode
  onClick?: (evt: Event) => void
}

export const footerProps = {
  cancel: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  cancelButton: Object as PropType<ButtonProps>,
  cancelLoading: Boolean,
  cancelText: String,
  cancelVisible: {
    type: Boolean,
    default: true,
  },
  footer: [Boolean, Array, Object] as PropType<boolean | FooterButtonProps[] | VNode>,
  ok: Function as PropType<(evt?: Event | unknown) => Promise<void> | void>,
  okButton: Object as PropType<ButtonProps>,
  okLoading: Boolean,
  okText: String,
} as const

export type FooterProps = ExtractInnerPropTypes<typeof footerProps>
export type FooterPublicProps = ExtractPublicPropTypes<typeof footerProps>
export type FooterComponent = DefineComponent<Omit<HTMLAttributes, keyof FooterPublicProps> & FooterPublicProps>
export type FooterInstance = InstanceType<DefineComponent<FooterProps>>
