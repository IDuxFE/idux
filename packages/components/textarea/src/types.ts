/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent, PropType, TextareaHTMLAttributes } from 'vue'

import { ɵInputCommonProps } from '@idux/components/input'

export const textareaProps = {
  ...ɵInputCommonProps,
  autoRows: { type: [Boolean, Object] as PropType<boolean | TextareaAutoRows>, default: undefined },
  computeCount: { type: Function as PropType<(value: string) => string>, default: undefined },
  maxCount: { type: [Number, String] as PropType<number | string>, default: undefined },
  resize: { type: String as PropType<TextareaResize>, default: undefined },
  showCount: { type: Boolean, default: undefined },
} as const

export type TextareaProps = ExtractInnerPropTypes<typeof textareaProps>
export type TextareaPublicProps = ExtractPublicPropTypes<typeof textareaProps>
export interface TextareaBindings {
  focus: (options?: FocusOptions) => void
  blur: () => void
}
export type TextareaComponent = DefineComponent<
  Omit<TextareaHTMLAttributes, keyof TextareaPublicProps> & TextareaPublicProps,
  TextareaBindings
>
export type TextareaInstance = InstanceType<DefineComponent<TextareaProps, TextareaBindings>>

export type TextareaResize = 'none' | 'both' | 'horizontal' | 'vertical'
export type TextareaAutoRows = { minRows?: number; maxRows?: number }
