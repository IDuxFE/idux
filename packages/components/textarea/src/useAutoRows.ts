/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TextareaAutoRows } from './types'
import type { ValueAccessor } from '@idux/cdk/forms'
import type { ComputedRef, Ref } from 'vue'

import { nextTick, onMounted, onUnmounted, watch, watchEffect } from 'vue'

import { isNumber, isObject, throttle } from 'lodash-es'

import { isFirefox } from '@idux/cdk/platform'
import { off, on, rAF } from '@idux/cdk/utils'

const isAutoRowsObject = (value: unknown): value is TextareaAutoRows => {
  return isObject(value) && isNumber(value.minRows) && isNumber(value.maxRows)
}

export function useAutoRows(
  textareaRef: Ref<HTMLTextAreaElement | undefined>,
  autoRows: ComputedRef<boolean | TextareaAutoRows>,
  accessor: ValueAccessor,
): void {
  let enabled = false
  let minRows: number | null = null
  let maxRows: number | null = null

  /** Keep track of the previous textarea value to avoid resizing when the value hasn't changed. */
  let previousValue: string | undefined
  let previousMinRows: number | null = null
  let initialHeight: string | undefined

  /** Cached height of a textarea with a single row. */
  let cachedLineHeight: number | undefined
  // padding + border
  const textareaBoxHeight = 10

  /** Class that should be applied to the textarea while it's being measured. */
  const measuringClass = isFirefox ? 'ix-textarea-autosize-measuring-firefox' : 'ix-textarea-autosize-measuring'

  let isDestroyed = false

  /**
   * Resize the textarea to fit its content.
   *
   * @param force Whether to force a height recalculation.
   * By default the height will be recalculated only if the value changed since the last call.
   */
  function resizeToFitContent(force = false) {
    // If autosizing is disabled, just skip everything else
    if (!enabled) {
      return
    }

    cacheTextareaLineHeight()

    // If we haven't determined the line-height yet,
    // we know we're still hidden and there's no point in checking the height of the textarea.
    if (!cachedLineHeight) {
      return
    }

    const textarea = textareaRef.value!
    const value = textarea.value

    // Only resize if the value or minRows have changed since these calculations can be expensive.
    if (!force && minRows === previousMinRows && value === previousValue) {
      return
    }

    const placeholderText = textarea.placeholder

    // Reset the textarea height to auto in order to shrink back to its default size.
    // Also temporarily force overflow:hidden, so scroll bars do not interfere with calculations.
    // Long placeholders that are wider than the textarea width may lead to a bigger scrollHeight
    // value. To ensure that the scrollHeight is not bigger than the content, the placeholders
    // need to be removed temporarily.
    textarea.classList.add(measuringClass)
    textarea.placeholder = ''

    const height =
      Math.round((textarea.scrollHeight - textareaBoxHeight) / cachedLineHeight) * cachedLineHeight + textareaBoxHeight

    // Use the scrollHeight to know how large the textarea *would* be if fit its entire value.
    textarea.style.height = `${height}px`
    textarea.classList.remove(measuringClass)
    textarea.placeholder = placeholderText

    nextTick(() => rAF(() => scrollToCaretPosition(textarea)))

    previousValue = value
    previousMinRows = minRows
  }

  /**
   * Cache the height of a single-row textarea if it has not already been cached.
   *
   * We need to know how large a single "row" of a textarea is in order to apply minRows and
   * maxRows. For the initial version, we will assume that the height of a single line in the
   * textarea does not ever change.
   */
  function cacheTextareaLineHeight(): void {
    if (cachedLineHeight) {
      return
    }

    // Use a clone element because we have to override some styles.
    const textarea = textareaRef.value!
    const textareaClone = textarea.cloneNode(false) as HTMLTextAreaElement
    textareaClone.rows = 1

    // Use `position: absolute` so that this doesn't cause a browser layout and use
    // `visibility: hidden` so that nothing is rendered. Clear any other styles that
    // would affect the height.
    textareaClone.style.position = 'absolute'
    textareaClone.style.visibility = 'hidden'
    textareaClone.style.border = 'none'
    textareaClone.style.padding = '0'
    textareaClone.style.height = ''
    textareaClone.style.minHeight = ''
    textareaClone.style.maxHeight = ''

    // In Firefox it happens that textarea elements are always bigger than the specified amount
    // of rows. This is because Firefox tries to add extra space for the horizontal scrollbar.
    // As a workaround that removes the extra space for the scrollbar, we can just set overflow
    // to hidden. This ensures that there is no invalid calculation of the line height.
    // See Firefox bug report: https://bugzilla.mozilla.org/show_bug.cgi?id=33654
    textareaClone.style.overflow = 'hidden'

    textarea.parentNode!.appendChild(textareaClone)
    cachedLineHeight = textareaClone.clientHeight
    textarea.parentNode!.removeChild(textareaClone)

    // Min and max heights have to be re-calculated if the cached line height changes
    setMinHeight()
    setMaxHeight()
  }

  /** Sets the minimum height of the textarea as determined by minRows. */
  function setMinHeight(): void {
    const minHeight = minRows && cachedLineHeight ? `${minRows * cachedLineHeight + textareaBoxHeight}px` : null

    if (minHeight) {
      textareaRef.value!.style.minHeight = minHeight
    }
  }

  /** Sets the maximum height of the textarea as determined by maxRows. */
  function setMaxHeight(): void {
    const maxHeight = maxRows && cachedLineHeight ? `${maxRows * cachedLineHeight + textareaBoxHeight}px` : null

    if (maxHeight) {
      textareaRef.value!.style.maxHeight = maxHeight
    }
  }

  /**
   * Scrolls a textarea to the caret position. On Firefox resizing the textarea will
   * prevent it from scrolling to the caret position. We need to re-set the selection
   * in order for it to scroll to the proper position.
   */
  function scrollToCaretPosition(textarea: HTMLTextAreaElement) {
    const { selectionStart, selectionEnd } = textarea

    // IE will throw an "Unspecified error" if we try to set the selection range after the
    // element has been removed from the DOM. Assert that the directive hasn't been destroyed
    // between the time we requested the animation frame and when it was executed.
    // Also note that we have to assert that the textarea is focused before we set the
    // selection range. Setting the selection range on a non-focused textarea will cause
    // it to receive focus on IE and Edge.
    if (!isDestroyed && document.activeElement === textarea) {
      textarea.setSelectionRange(selectionStart, selectionEnd)
    }
  }

  function reset() {
    // Do not try to change the textarea, if the initialHeight has not been determined yet
    // This might potentially remove styles when reset() is called before ngAfterViewInit
    if (initialHeight !== undefined) {
      textareaRef.value!.style.height = initialHeight
    }
  }

  const onResize = throttle(() => resizeToFitContent(true), 16)

  onMounted(() => {
    // Remember the height which we started with in case autosizing is disabled
    initialHeight = textareaRef.value!.style.height

    watchEffect(() => {
      const _autoRows = autoRows.value
      if (isAutoRowsObject(_autoRows)) {
        enabled = true
        minRows = _autoRows.minRows
        maxRows = _autoRows.maxRows
      } else {
        enabled = _autoRows
        minRows = null
        maxRows = null
      }
      enabled ? resizeToFitContent(true) : reset()
    })

    watch(
      () => accessor.valueRef.value,
      () => resizeToFitContent(),
    )

    on(window, 'resize', onResize)
  })

  onUnmounted(() => {
    off(window, 'resize', onResize)
    isDestroyed = true
  })
}
