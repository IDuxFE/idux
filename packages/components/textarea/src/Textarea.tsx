import type { ComputedRef, Ref, Slot } from 'vue'
import type { ValueAccessor } from '@idux/cdk/forms'
import type { TextareaConfig } from '@idux/components/config'
import type { TextareaProps } from './types'

import { computed, defineComponent } from 'vue'
import { normalizeClass } from '@vue/shared'
import { useGlobalConfig } from '@idux/components/config'
import { IxIcon } from '@idux/components/icon'
import { ɵUseCommonBindings } from '@idux/components/input'
import { useAutoRows } from './useAutoRows'
import { textareaProps } from './types'

export default defineComponent({
  name: 'IxTextarea',
  props: textareaProps,
  setup(props, { slots, expose, attrs }) {
    const config = useGlobalConfig('textarea')

    const {
      elementRef,
      valueAccessor,

      isDisabled,
      clearIcon,
      clearHidden,
      isClearable,
      isFocused,

      focus,
      blur,

      handlerInput,
      handlerCompositionStart,
      handlerCompositionEnd,
      handlerFocus,
      handlerBlur,
      handlerClear,
    } = ɵUseCommonBindings(props, config)

    expose({ focus, blur })

    const classes = useClasses(props, config, isFocused, isDisabled)
    const dataCount = useDataCount(props, config, valueAccessor)
    const autoRows = computed(() => props.autoRows ?? config.autoRows)
    const resize = computed(() => {
      let resize = props.resize ?? config.resize
      if (autoRows.value) {
        return resize === 'horizontal' ? 'horizontal' : 'none'
      } else {
        return resize
      }
    })
    const textareaStyle = computed(() => ({ resize: resize.value }))

    useAutoRows(elementRef as Ref<HTMLTextAreaElement>, autoRows, valueAccessor)

    return () => {
      const suffix = renderSuffix(isClearable.value, slots.clearIcon, clearIcon.value, clearHidden.value, handlerClear)
      const { class: className, style, ...rest } = attrs
      return (
        <span class={normalizeClass([classes.value, className])} style={style as any} data-count={dataCount.value}>
          <textarea
            {...rest}
            ref={elementRef}
            class="ix-textarea-inner"
            style={textareaStyle.value}
            disabled={isDisabled.value}
            readonly={props.readonly}
            onInput={handlerInput}
            onCompositionstart={handlerCompositionStart}
            onCompositionend={handlerCompositionEnd}
            onFocus={handlerFocus}
            onBlur={handlerBlur}
          ></textarea>
          {suffix}
        </span>
      )
    }
  },
})

function useClasses(
  props: TextareaProps,
  config: TextareaConfig,
  isFocused: Ref<boolean>,
  disabled: ComputedRef<boolean>,
) {
  return computed(() => {
    const sizeClass = `ix-textarea-${props.size ?? config.size}`
    return {
      'ix-textarea': true,
      [sizeClass]: true,
      'ix-textarea-disabled': disabled.value,
      'ix-textarea-focused': isFocused.value,
      'ix-textarea-with-count': props.showCount ?? config.showCount,
    }
  })
}

function useDataCount(props: TextareaProps, config: TextareaConfig, valueAccessor: ValueAccessor) {
  return computed(() => {
    const showCount = props.showCount ?? config.showCount
    const computeCount = props.computeCount ?? config.computeCount
    const maxCount = props.maxCount ?? config.maxCount
    let dataCount = ''
    if (showCount) {
      const value = valueAccessor.value ?? ''
      dataCount = value.length
      if (computeCount) {
        dataCount = computeCount(value)
      } else if (maxCount) {
        dataCount += ' / ' + maxCount
      }
    }
    return dataCount
  })
}

function renderSuffix(
  isClearable: boolean,
  clearIconSlot: Slot | undefined,
  clearIcon: string,
  clearHidden: boolean,
  handlerClear: (evt: MouseEvent) => void,
) {
  if (!isClearable) {
    return null
  }

  const classes = { 'ix-textarea-suffix': true, 'ix-textarea-suffix-hidden': clearHidden }
  const child = clearIconSlot?.({ handlerClear }) ?? <IxIcon name={clearIcon} onClick={handlerClear}></IxIcon>
  return <span class={classes}>{child}</span>
}
