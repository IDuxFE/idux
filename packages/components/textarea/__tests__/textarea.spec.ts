import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import IxTextarea, { renderSuffix } from '../src/Textarea'
import { TextareaProps } from '../src/types'

describe('Textarea', () => {
  const TextareaMount = (options?: MountingOptions<Partial<TextareaProps>>) => mount(IxTextarea, { ...options })

  renderWork(IxTextarea)

  test('v-model work', async () => {
    const onUpdateValue = vi.fn()
    const wrapper = TextareaMount({ props: { value: 'init value', 'onUpdate:value': onUpdateValue } })
    const textarea = wrapper.find('textarea')

    expect(textarea.element.value).toBe('init value')

    await textarea.setValue('input setValue')

    expect(onUpdateValue).toBeCalledTimes(1)
    expect(onUpdateValue).toBeCalledWith('input setValue')

    await wrapper.setProps({ value: 'wrapper setProps' })

    expect(textarea.element.value).toBe('wrapper setProps')

    textarea.element.value = '使用拼音'
    await textarea.trigger('compositionstart')

    expect(wrapper.emitted()).toHaveProperty('compositionstart')
    expect(onUpdateValue).toBeCalledTimes(1)

    await textarea.trigger('input')

    expect(wrapper.emitted()).toHaveProperty('input')
    expect(onUpdateValue).toBeCalledTimes(1)

    await textarea.trigger('compositionend')

    expect(wrapper.emitted()).toHaveProperty('compositionend')
    expect(onUpdateValue).toBeCalledTimes(2)
    expect(onUpdateValue).toBeCalledWith('使用拼音')
  })

  test('disabled work', async () => {
    const onFocus = vi.fn()
    const onBlur = vi.fn()
    const wrapper = TextareaMount({ props: { disabled: true, onFocus, onBlur } })
    //await wrapper.find('textarea').trigger('focus')

    expect(wrapper.classes()).toContain('ix-textarea-disabled')
    // expect(wrapper.classes()).not.toContain('ix-textarea-focused')
    // expect(onFocus).not.toBeCalled()

    // await wrapper.find('textarea').trigger('blur')

    // expect(onBlur).not.toBeCalled()

    await wrapper.setProps({ disabled: false })
    //await wrapper.find('textarea').trigger('focus')

    expect(wrapper.classes()).not.toContain('ix-textarea-disabled')
    //expect(wrapper.classes()).toContain('ix-textarea-focused')
    //expect(onFocus).toBeCalled()

    // await wrapper.find('textarea').trigger('blur')

    // expect(onBlur).toBeCalled()
  })

  // test('readonly work', async () => {
  //   const onFocus = vi.fn()
  //   const onBlur = vi.fn()
  //   const wrapper = TextareaMount({ props: { readonly: true, onFocus, onBlur } })
  //   await wrapper.find('textarea').trigger('focus')

  //   expect(onFocus).toBeCalled()

  //   await wrapper.find('textarea').trigger('blur')

  //   expect(onBlur).toBeCalled()
  // })

  test('size work', async () => {
    const wrapper = TextareaMount({ props: { size: 'lg' } })

    expect(wrapper.classes()).toContain('ix-textarea-lg')

    await wrapper.setProps({ size: 'sm' })
    expect(wrapper.classes()).toContain('ix-textarea-sm')

    await wrapper.setProps({ size: undefined })
    expect(wrapper.classes()).toContain('ix-textarea-md')
  })

  test('clearable work', async () => {
    const onClear = vi.fn()
    const wrapper = TextareaMount({ props: { clearable: true, onClear } })
    const renderSuffixFn = vi.fn(renderSuffix)

    expect(wrapper.find('.ix-icon-close-circle').exists()).toBe(true)
    expect(wrapper.find('.ix-textarea-suffix-hidden').exists()).toBe(true)

    await wrapper.find('.ix-icon-close-circle').trigger('click')

    expect(onClear).toBeCalled()

    await wrapper.setProps({ value: 'value' })

    expect(wrapper.find('.ix-textarea-suffix-hidden').exists()).toBe(false)

    const res = (renderSuffixFn(true, undefined, '', true, onClear, '', true)?.props as { class: '' }).class

    expect(res).toBe('-suffix -suffix-scroll')

    await wrapper.setProps({ disabled: true })

    expect(wrapper.find('.ix-textarea-suffix-hidden').exists()).toBe(true)

    await wrapper.setProps({ disabled: false, readonly: true })

    expect(wrapper.find('.ix-textarea-suffix-hidden').exists()).toBe(true)

    await wrapper.setProps({ clearable: false })

    expect(wrapper.find('.ix-icon-close-circle').exists()).toBe(false)
  })

  test('resize and autoRows work', async () => {
    const wrapper = TextareaMount({ props: { resize: 'none' } })

    const textarea = wrapper.find('textarea')
    let previousHeight = textarea.element.clientHeight

    expect(textarea.element.style.resize).toBe('none')

    await wrapper.setProps({ resize: 'both' })

    expect(textarea.element.style.resize).toBe('both')

    await wrapper.setProps({ resize: 'horizontal' })

    expect(textarea.element.style.resize).toBe('horizontal')

    await wrapper.setProps({ autoRows: true })

    await wrapper.setProps({ resize: 'vertical' })
    expect(textarea.element.style.resize).toBe('none')

    expect(textarea.element.clientHeight).toBe(previousHeight)

    await wrapper.setProps({ value: `1\n2\n3\n4\n` })

    // TODO: switch to e2e, jest doesn't do inline styles
    // expect(textarea.element.clientHeight).toBeGreaterThan(previousHeight)
    // expect(textarea.element.clientHeight).toBe(textarea.element.scrollHeight)
    expect(wrapper.html()).toMatchSnapshot()

    previousHeight = textarea.element.clientHeight

    await wrapper.setProps({ value: `1\n2\n3\n4\n5\n6\n` })

    // jest doesn't do inline styles
    // expect(textarea.element.clientHeight).toBeGreaterThan(previousHeight)
    // expect(textarea.element.clientHeight).toBe(textarea.element.scrollHeight)
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ autoRows: { minRows: 2, maxRows: 4 } })

    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ value: `1\n` })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('showCount, maxCount and computeCount  work', async () => {
    const wrapper = TextareaMount({ props: { showCount: true } })

    expect(wrapper.classes()).toContain('ix-textarea-with-count')
    expect(wrapper.attributes('data-count')).toBe('0')

    await wrapper.setProps({ value: 'test' })

    expect(wrapper.attributes('data-count')).toBe('4')

    await wrapper.setProps({ maxCount: '100' })

    expect(wrapper.attributes('data-count')).toBe('4 / 100')

    await wrapper.setProps({ computeCount: (value: string) => `${value.length * 2} - 100` })

    expect(wrapper.attributes('data-count')).toBe('8 - 100')
  })
})
