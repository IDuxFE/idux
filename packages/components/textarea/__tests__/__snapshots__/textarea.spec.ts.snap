// Vitest Snapshot v1

exports[`Textarea > render work 1`] = `"<span class=\\"ix-textarea ix-textarea-md\\" data-count=\\"\\"><textarea class=\\"ix-textarea-inner\\" style=\\"resize: vertical;\\"></textarea><!----></span>"`;

exports[`Textarea > resize and autoRows work 1`] = `"<span class=\\"ix-textarea ix-textarea-md\\" data-count=\\"\\"><textarea class=\\"ix-textarea-inner\\" style=\\"resize: none;\\"></textarea><!----></span>"`;

exports[`Textarea > resize and autoRows work 2`] = `"<span class=\\"ix-textarea ix-textarea-md\\" data-count=\\"\\"><textarea class=\\"ix-textarea-inner\\" style=\\"resize: none;\\"></textarea><!----></span>"`;

exports[`Textarea > resize and autoRows work 3`] = `"<span class=\\"ix-textarea ix-textarea-md\\" data-count=\\"\\"><textarea class=\\"ix-textarea-inner\\" style=\\"resize: none;\\"></textarea><!----></span>"`;

exports[`Textarea > resize and autoRows work 4`] = `"<span class=\\"ix-textarea ix-textarea-md\\" data-count=\\"\\"><textarea class=\\"ix-textarea-inner\\" style=\\"resize: none;\\"></textarea><!----></span>"`;
